### PR TITLE
Add library propagation rules to control detailed sample library conf…

### PR DESCRIPTION
**Pull request by apmasell**
…igurations

### CODE REVIEW

File: miso-web/src/main/webapp/pages/editLibrary.jsp [hunk 1]
```diff
@@ -1453,6 +1459,7 @@ function submitBulkLibraries() {
 </div>
 
 <script type="text/javascript">
+  propagationRules = ${propagationRulesJSON};
   jQuery(document).ready(function () {
```
Comment (callunity) [inline miso-web/src/main/webapp/pages/editLibrary.jsp]: Can we put this under the Library namespace?

Comment (apmasell) [inline miso-web/src/main/webapp/pages/editLibrary.jsp]: Done.

**Comment by callunity** [comment]
:+1:

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/manager/MisoRequestManager.java [hunk 2]
```diff
@@ -1642,7 +1648,13 @@ public long saveEmPcrDilution(emPCRDilution dilution) throws IOException {
   @Override
   public long saveLibrary(Library library) throws IOException {
     if (libraryStore != null) {
-      return libraryStore.save(library);
+      SampleAdditionalInfo info = library.getSample().getSampleAdditionalInfo();
+      if (LibraryPropagationRule.validate(library,
+          libraryPropagationRuleDao.getLibraryPropagationRulesByClass(info == null ? null : info.getSampleClass()))) {
+        return libraryStore.save(library);
+      } else {
+        throw new IOException("Invalid propagation.");
```
Comment (tonydebat) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/manager/MisoRequestManager.java]: When using the single sample/library ui to generate a library from a detailed sample a stack trace is displayed when the propagation validation fails.

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/LibraryPropagationRule.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  public void setLibraryStrategyType(Long libraryStrategyType) {
+    this.libraryStrategyType = libraryStrategyType;
+  }
+
+  public boolean validate(Library library) {
+    if (library.getSample().getSampleAdditionalInfo() == null) return true;
+    if (sampleClass != null
+        && library.getSample().getSampleAdditionalInfo().getSampleClass().getSampleClassId() != sampleClass.getSampleClassId())
+      return false;
+    if (platformName != null && !library.getPlatformName().equals(platformName)) return false;
```
Comment (tonydebat) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/LibraryPropagationRule.java]: This line should be:
if (platformName != null && !library.getPlatformName().equals(platformName**.getKey()**)) return false;

Comment (apmasell) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/LibraryPropagationRule.java]: Fixed.

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/manager/MisoRequestManager.java (hunk 2 above)
Comment (apmasell) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/manager/MisoRequestManager.java]: Fixed.

**Comment by tonydebat** [comment]
I reopened https://jira.oicr.on.ca/browse/GLT-523 to describe another UI error getting.
:+1:

**Pull request merged**